builtin/umask.c

```diff
@@ -7,21 +7,146 @@
 #include <sys/stat.h>
 #include "builtin.h"
 
-/*
- * TODO: Implement symbolic mode
- */
-
 static const char umask_usage[] = "usage: umask [-S] [mode]\n";
 
+enum umask_symbolic_state {
+	UMASK_WHO,
+	UMASK_PERM,
+};
+
+static mode_t umask_current_mask(void) {
+	mode_t mask;
+	const mode_t default_mode = 0022;
+
+	mask = umask(default_mode);
+	umask(mask);
+	return mask;
+}
+
+static int umask_update_mode(mode_t *mode, char action, mode_t *perm_mask, mode_t *who_mask) {
+	switch (action) {
+	case '+':
+		*mode |= (*who_mask & *perm_mask);
+		break;
+
+	case '=':
+		*mode = (*mode & ~(*who_mask)) | (*perm_mask & *who_mask);
+		break;
+
+	case '-':
+		*mode &= (0777 & ~(*who_mask & *perm_mask));
+		break;
+
+	default:
+		fprintf(stderr, "unknown action -- '%c'\n", action);
+		return EXIT_FAILURE;
+		break;
+	}
+
+	*perm_mask = *who_mask = 0;
+	return EXIT_SUCCESS;
+}
+
+static int umask_mode(mode_t *mode, char *symbolic) {
+	mode_t tmp_mode = 0777 & ~(umask_current_mask());
+	char *c = symbolic;
+	enum umask_symbolic_state state = UMASK_WHO;
+	mode_t who_mask = 0;
+	mode_t perm_mask = 0;
+	char action = '\0';
+
+
+	for (; *c != '\0'; c++) {
+		switch(state) {
+		case UMASK_WHO:
+			switch(*c) {
+			case 'u':
+				who_mask |= S_IRWXU;
+				break;
+			case 'g':
+				who_mask |= S_IRWXG;
+				break;
+			case 'o':
+				who_mask |= S_IRWXO;
+				break;
+			case 'a':
+				who_mask |= (S_IRWXU | S_IRWXG | S_IRWXO);
+				break;
+			case '+':
+			case '-':
+			case '=':
+				if (who_mask == 0) {
+					who_mask |= (S_IRWXU | S_IRWXG | S_IRWXO);
+				}
+				action = *c;
+				state = UMASK_PERM;
+				break;
+			default:
+				fprintf(stderr, "Unknown who -- '%c'\n", *c);
+				return EXIT_FAILURE;
+			}
+
+			break;
+
+		case UMASK_PERM:
+			switch (*c) {
+			case 'u':
+				perm_mask = (tmp_mode & S_IRWXU) | ((tmp_mode & S_IRWXU) >> 3) | ((tmp_mode & S_IRWXU) >> 6);
+				break;
+			case 'g':
+				perm_mask = ((tmp_mode & S_IRWXG) << 3) | (tmp_mode & S_IRWXG) | ((tmp_mode & S_IRWXG) >> 3);
+				break;
+			case 'o':
+				perm_mask = ((tmp_mode & S_IRWXO) << 6) | ((tmp_mode & S_IRWXO) << 3) | (tmp_mode & S_IRWXO);
+				break;
+			case 'r':
+				perm_mask |= (S_IRUSR | S_IRGRP | S_IROTH);
+				break;
+			case 'w':
+				perm_mask |= (S_IWUSR | S_IWGRP | S_IWOTH);
+				break;
+			case 'x':
+				perm_mask |= (S_IXUSR | S_IXGRP | S_IXOTH);
+				break;
+			case ',':
+				state = UMASK_WHO;
+				if (umask_update_mode(&tmp_mode, action, &perm_mask, &who_mask) != EXIT_SUCCESS) {
+					return EXIT_FAILURE;
+				}
+				break;
+			default:
+				fprintf(stderr, "Invalid permission -- '%c'\n", *c);
+				return EXIT_FAILURE;
+			}
+			break;
+		}
+	}
+
+	if (state == UMASK_PERM) {
+		if (umask_update_mode(&tmp_mode, action, &perm_mask, &who_mask) != EXIT_SUCCESS) {
+			return EXIT_FAILURE;
+		}
+	} else {
+		fprintf(stderr, "Unexpected state %d\n", state);
```

Comment:
This error message could be simplified to something like "missing permissions from symbolic mode".